user/man/evlogmgr.1
."Generated by db2man.xsl. Don't modify this, modify the source.
.de Sh \" Subsection
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Ip \" List item
.br
.ie \\n(.$>=3 .ne \\$3
.el .ne 3
.IP "\\$1" \\$2
..
.TH "evlogmgr"  "" "" ""
.SH NAME
evlogmgr \- Event log manager
.SH "SYNOPSIS"
\fBevlogmgr\fR   \fB-c\fR | \fB--compact  \fIfilter\fR \fR 
 [ \fB-F\fR | \fB--force\fR ] [ \fB-C\fR | \fB--compr-bak\fR ] 
 [[\fB \fB-p\fR | \fB--private\fR \fR ] | [\fB \fB-l\fR | \fB--log  \fIsrcfile\fR \fR \fR]]

\fBevlogmgr\fR   \fB-f\fR | \fB--fix\fR  
 [[\fB \fB-p\fR | \fB--private\fR \fR ] | [\fB \fB-l\fR | \fB--log  \fIsrcfile\fR \fR \fR]]

\fBevlogmgr\fR   \fB-s\fR | \fB--show-status  \fIfilter\fR \fR  
 [[\fB \fB-p\fR | \fB--private\fR \fR ] | [\fB \fB-l\fR | \fB--log  \fIsrcfile\fR \fR \fR]]
.SH "DESCRIPTION"

.PP
The \fBevlogmgr\fR command performs log management on the event log, on the private log, or optionally, on a log file that you specify. You also specify which events are to be deleted. The space freed by deleted events is reused for undeleted events (a process referred to as compaction) and the log file is truncated, thus reducing its overall size.

.PP
You must have root permission to use this command.

.SH "OPTIONS"

.TP
-\fBc\fR, --\fBcompact\fR \fIfilter\fR 
Deletes event records in the event log matching the specified filter, fills space freed by deleting the event records (by moving later records into the free space), and then truncates the log thus establishing a new write offset for new events. The sequential order of events in the event log is maintained during this process and the record ids of the remaining event records are left unchanged. See the \fBevlquery\fR man page for more information about the \fIfilter\fR.

Before the compaction process begins, a backup copy of the log file is created if space is available. Following successful compaction, the backup copy is deleted. If the \fBevlogmgr\fR command is issued from the command line, and the user types Ctrl-C, then the active log file is restored from the backup copy. In the event that the system reboots before compaction completes, the active log file is restored from the backup copy (after reboot).

Unless the --\fBprivate\fR or --\fBlog\fR options are used, log management is performed on \fI/var/evlog/eventlog\fR.

The --\fBforce\fR option should only be used as a last resort. If space is not available to create a backup (compressed or uncompressed), then this option forces compaction to occur without first creating a backup of the log file. If compaction does not complete successfully, then the \fBevlogmgr\fR command with the --\fBfix\fR option (see below) might be able to restore the log file to a usable state.

The --\fBcompr-bak\fR option specifies that the backup log file created prior to compacting the original log file should be compressed. This option will take longer, but limited space might require its use.

The --\fBprivate\fR option specifies that compaction should be performed on \fI/var/evlog/privatelog\fR. This option is not valid in combination with the --\fBlog\fR option.

The --\fBlog\fR option specifies an alternate log file, \fIsrcfile\fR, to be compacted. This option is not valid in combination with the --\fBprivate\fR option.

.TP
-\fBf\fR, --\fBfix\fR 
This option is used to repair a corrupted log file, and is required only in cases where the --\fBforce\fR option was used and the log management tasks terminated before completion.

Each record in the event log is validated, and any records that are deemed invalid are deleted and replaced by valid records in the log. The command reports the total number of bytes deleted.

Some of the error messages that will indicate the need to repair the log include:

\(bu Input/output error (from evlview)
.Pp

\(bu Failed to count record for deletion 
    (from evlogmgr -s)
.Pp

\(bu Failed to compact log  (from evlogmgr -c)
.Pp

Unless the --\fBprivate\fR or --\fBlog\fR options are used, repair is performed on \fI/var/evlog/eventlog\fR.

The --\fBprivate\fR option specifies that repair should be performed on \fI/var/evlog/privatelog\fR. This option is not valid in combination with the --\fBlog\fR option.

The --\fBlog\fR option specifies an alternate log file, \fIsrcfile\fR, to be repaired. This option is not valid in combination with the --\fBprivate\fR option.

.TP
-\fBs\fR, --\fBshow-status\fR \fIfilter\fR 
Shows how many records in the event log match the \fIfilter\fR, and shows the total number of records in the log. See the \fBevlquery\fR man page for more information about the \fIfilter\fR.

Unless the --\fBprivate\fR or --\fBlog\fR options are used, status is shown for \fI/var/evlog/eventlog\fR.

The --\fBprivate\fR option specifies that status should be shown for \fI/var/evlog/privatelog\fR. This option is not valid in combination with the --\fBlog\fR option.

The --\fBlog\fR option specifies that status should be shown for an alternate log file, \fIsrcfile\fR. This option is not valid in combination with the --\fBprivate\fR option.

.SH "EXAMPLES"

.TP
Example 1
Run \fBevlogmgr\fR -\fBc\fR from a cron job. See "man 5 crontab" for more details.

# Daily at 2 AM remove recds older than 30 days

0 2 * * * root /sbin/evlogmgr -c 'age > "30d"'

      

# Daily at 1 AM remove recds w/ sev=DEBUG

0 1 * * * root /sbin/evlogmgr -c "severity=DEBUG"

.TP
Example 2
Issuing the following command:

   evlogmgr -c 'severity=DEBUG' -C 
   
might result in the following events being appended to the log:

recid=589, size=74, format=STRING, event_type=0x2, facility=LOGMGMT, 
severity=NOTICE, uid=root, gid=root, pid=18653, pgrp=18653, 
time=Thu 09 May 2002 12:59:08 PM PDT, flags=0x0, thread=0x400, processor=0
Log compaction on /var/evlog/eventlog starts at Thu May 09 12:59:08 2002
          
recid=591, size=69, format=STRING, event_type=0x3, facility=LOGMGMT, 
severity=NOTICE, uid=root, gid=root, pid=18653, pgrp=18653, 
time=Thu 09 May 2002 12:59:09 PM PDT, flags=0x0, thread=0x400, processor=0
Log compaction on /var/evlog/eventlog ended. 48 events were removed. 
.TP
Example 3
Issuing the following command:

   evlogmgr -s 'severity=DEBUG'
 
might display the following:

   Total number of records is 294.
   Number of records matching the filter is 48.
   Log file size would be reduced by 5.00 kbytes
.TP
Example 4
Issuing the following command:

   evlogmgr --fix
  
might display the following:

  Found a corrupted event record. Trying to resync with next good record...
  Log repair finished, problem fixed.  Discarded 6 bytes. 
.SH "FILES"

.TP
\fI/var/evlog/eventlog\fR
standard event log

.TP
\fI/var/evlog/privatelog\fR
private event log

.SH "SEE ALSO"
\fIevlquery man page\fR    Filter expression syntax rules



CHANGES
Enterprise-Class Event Logging for Linux
========================================

v1.0.0   9/20/2001
Initial  Version


v1.0.1  10/01/2001
- Replaced calls to spin_lock_irq() and spin_unlock_irq() with calls
  to  spin_lock_irqsave() and spin_unlock_irqrestore() in kernel log
  write functions.  This resolved a hang condition experienced with 
  a SCSI device driver.
- Replaced use of sockaddr to sockaddr_un to provide longer socket 
  path name.
- Several minor changes needed for s390 (32 and 64 bit).
- runtests.sh scripts modifed to take option of s390 (32-bit) or
  s390x (64-bit) to specify s390/IBM Z-Series.
- Processing of white-spaces was fixed to correctly process >1 white 
  space.
NOTE:  The patch for GNU C Library, evlog-1.0.1_glibc2.2.3.patch,
is unchanged from version 1.0.0.


v1.1.0  11/09/2001

New features and enhancements in v1.1.0
    
1. Kernel patches for kernel versions 2.4.2 through 2.4.14 will 
   now be posted on the website.  
2. A new library that "wraps" around GNU C Lib and provides
   forwarding of syslog messages to the POSIX event log without
   having to patch and rebuild glibc.
3. Formatting Template support and evltc command (for compiling
   formatting template). Some features are not yet available
   (section numbers in the spec are indicated):
* (6.1.2.2.1.2: Imports and Typedefs) If an import statement specifies
   a struct that has already been created or imported, the import
   statement will be silently ignored.
* (6.1.2.2.1.4: Attributes Section [overall format]) The issue regarding
   const-only structs is still unresolved.  It is currently impossible 
   to define a struct that has only const attributes.
* (6.1.2.2.1.4.3: Attribute Dimension or Bit-Field Width) Bit fields 
   and arrays of structs are not yet supported.
* (6.1.2.2.1.4.5 Attribute Format)   The following is not yet supported:
* Validation of printf-style formats is incomplete.  In particular, you
  are permitted to specify formats that are completely inappropriate for
  the given data type -- e.g. "%f" for a string.
* The %t (dump) format is currently supported only for arrays.
* %b (bitmap) and %v (named-value) specifiers are not yet supported
  for arrays.
* As previously mentioned, arrays of structs are not yet supported;
  so the %Z format specifier is not, either.
* (6.2.3.3 Formatting Templates [API]) The following is not supported:
   * The evl_populatetemplate() function does not currently create the 
     _EXTRA_DATA_ pseudo-attribute
   * The evltemplate_getatt() function does not currently handle names 
     such as "x.y" to reference members of struct-type attributes.

4. Displaying of events logged in the kernel (with the new log write
   functions) on the system console.  The --output option was added to 
   the evlconfig command to turn console logging on/off, or set the
   severity level for displaying to the console.
5. Forwarding of printk() messages to the kernel event buffer must now
   be explicitly enabled when configuring the kernel (forwarding is
   disabled by default).
6. evlfacility command for managing the facility registry.
7. Restricted Logging support. 
8. Added "address" as a data type for formatting templates.
9. Added 2 new purpose arguments for posix_log_query_create() as
   extensions to the draft POSIX standard:  EVL_PRPS_TEMPLATE and
   EVL_PRPS_RESTRICTED.
10. Added EVL_KERNEL_EVENT flag for log_flags member in 
    posix_log_entry.
11. Added varargs versions of log write functions in kernel and 
    user-space.
12. Added support for --new and --timeout options in evlview
    command.
13. Added --templates and --notemplates options to evlview command.
14. Added unlink(PidFile) function in evlogd.c, evlnotifyd.c,
    evlactiond.c.
15. Moved evlogd, evlnotifyd, and evlactiond pid files from /var/evlog
    to /var/run.

Bug Fixes in v1.1.0

1. Removed evlog *.pid files on graceful shutdown or on startup if 
   previous shutdown was non-graceful to fix evlogd daemon startup
   failure.
2. In event notification, fixed a bug where error returns from
   rt_sigqueueinfo() were causing the notification to be erroneously
   disabled.
3. In do_syslog(), now use options 20 and 21 for read/write to kernel
   event buffer, instead of 9 and 10, to avoid conflict with other
   newly added options in do_syslog(). 
4. posix_log_notify_add() is now signal-safe.


v1.1.1  11/28/2001

Bug fixes in v1.1.1

1. (Fix for Bug # 485676) Code added in v1.1.0 for displaying events on the
   system console is incompatible with the re-write of code for displaying of
   messages from printk() first appearing in kernel v2.4.10. This causes the
   kernel to hang or PANIC under certain conditions.  This only affects the
   patches for kernel versions 2.4.10 through 2.4.14. 
2. Added check in the kernel version of posix_log_write() for null-terminated
   string (when log_format is POSIX_LOG_STRING).  Returns EBADMSG if string
   is not null-terminated.  A few other changes to make this function   
   consistent between kernel and user space.
3. Fixed bug causing programs linked with libevl.so to fail at link time.
4. Added missing function call in libevlsyslog.so so that syslog messages
   actually get logged in event log.
5. Fixed several bugs causing problems on s390 and/or 64-bit machines:
   * evltc command reporting bogus errors or causing segfaults
   * Backward seeks through the event log (as used by evlview --tail) failed. 
   * Various memory corruption problems 
6. Updated all header files for C++ compatability.


v1.2.0  02/18/2002

New Features and Enhancements in v1.2.0

1. Events logged in user-space are now passed directly to the evlogd 
   daemon,  instead of being written to the kernel event buffer.
2. The libevl event logging library no longer depends on the event 
   logging kernel patch, meaning that event logging in user-space 	
   does not require  changes in the kernel.  The kernel patch is 

user/cmd/evlogmgr/evlogmgr.c
/*
 * Linux Event Logging for the Enterprise
 * Copyright (c) International Business Machines Corp., 2002
 *
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 *
 *  Please send e-mail to lkessler@users.sourceforge.net if you have
 *  questions or comments.
 *
 *  Project Website:  http://evlog.sourceforge.net/
 *
 */
#include <stdio.h>
#include <fcntl.h>
#include <signal.h>
#include <sys/param.h>
#include <sys/types.h>
#include <errno.h>
#include <stdlib.h>
#include <unistd.h>
#include <poll.h>
#include <values.h>
#include <unistd.h>
#include <string.h>
#include <strings.h>		/* Defines bcopy() */
#include <sys/stat.h>
#include <time.h>
#include <assert.h>

#include "config.h"
#include "posix_evlog.h"
#include "posix_evlsup.h"
#include "evlog.h"
#include "evl_logmgmt.h"

#ifdef DEBUG2
#define TRACE(fmt, args...)		fprintf(stdout, fmt, ##args)
#else
#define TRACE(fmt, args...)		/* fprintf(stdout, fmt, ##args) */
#endif


#define BOOLARG 	1
#define STRINGARG 	2
#define INTARG 		3

char *progName = 0;
char *logfile;
/* String args */
char *cfilterStr = NULL;
char *showstat = NULL;
char *logFilePath = NULL;


int privFlag = 0;
int forceFlag = 0;

int repairFlag = 0;
int helpFlag = 0;
int somethingBesidesHelp = 0;

int gzBackUpFlag = 0;

struct argInfo {
	char	*argName;
	char 	*shrtName;
	void	*argVal;
	int	argType;
	int	argSeen;
};

struct argInfo argInfo[] = {
	{ "--private",		"-p", 		&privFlag, 			BOOLARG, 	0},	
	{ "--force",		"-F", 		&forceFlag, 		BOOLARG, 	0},	
	{ "--show-status",	"-s", 		&showstat, 			STRINGARG, 	0},	
	{ "--compact",		"-c",		&cfilterStr,		STRINGARG,	0},
	{ "--fix",			"-f", 		&repairFlag, 		BOOLARG, 	0},	
	{ "--log",			"-l", 		&logFilePath, 		STRINGARG, 	0},	
	{ "--compr-bak",	"-C",		&gzBackUpFlag, 		BOOLARG,	0},
	{ "--help",			"-h",		&helpFlag, 			BOOLARG,	0},
	{ NULL,				NULL,		NULL,			0,			0}
};

void usage()
{
	fprintf(stderr, "Usage:\n"
					"evlogmgr  -c | --compact <filter> [-F | --force] [-C | compr-bak]\n"
					"         [[-p | --private] | [-l | --log <logfilepath>]]\n"
					"evlogmgr  -f | --fix\n" 
					"         [[-p | --private] | [-l | --log <logfilepath>]]\n"
					"evlogmgr  -s | --show-status <filter>\n"
					"         [[-p | --private] | [-l | --log <logfilepath>]]\n");
	exit(1);
}

struct argInfo *
findArg(const char *argName)
{
	struct argInfo *ai;
	for (ai = argInfo; ai->argName; ai++) {
		if (!strcmp(argName, ai->argName) || !strcmp(argName, ai->shrtName)) {
			return ai;
		}
	}
	return NULL;
}

/*
 * Processes the arg list and stores the values of the various options in
 * the appropriate variables.  Detects duplicate args, args without values,
 * etc.  No real semantic checks -- e.g., -timeout without -new.
 */
void
processArgs(int argc, char **argv)
{
	struct argInfo *ai;
	int n;

	for (++argv; *argv; ++argv) {
		char *arg = *argv;
		ai = findArg(arg);
		if (!ai) {
			fprintf(stderr, "%s: Unknown option: %s\n", progName, arg);
			usage();
		}

		if (ai->argSeen) {
			fprintf(stderr,"%s: Multiple %s options not supported\n",
				progName, arg);
			exit(1);
		}
		ai->argSeen = 1;

		if (strcmp(arg, "--help") && strcmp(arg, "-h")) {
			somethingBesidesHelp = 1;
		}

		if (ai->argType != BOOLARG) {
			argv++;
			arg = *argv;
			if (!arg) {
				fprintf(stderr,
					"%s: Missing value for %s option\n",
					progName, ai->argName);
				usage();
			}
		}

		switch (ai->argType) {
    		case BOOLARG:
    		{
    			*((int*)(ai->argVal)) = 1;
    			break;
    		}
    		case STRINGARG:
    		{
    			char *endChar = 0;
    	
    			*((char**)(ai->argVal)) = arg;
    			break;
    		}
    		case INTARG:
    		{
    			char *endChar = 0;
    		    long val = strtol(arg, &endChar, 0);
    			if (*endChar != '\0' || val == -1) {
    				fprintf(stderr,
    					"%s: Illegal value for %s option\n",
    					progName, ai->argName);
    				usage();
    			}
    			*((int*)(ai->argVal)) = (int) val;
    		    break;
    		}
    		default:
    			fprintf(stderr,"%s: Internal error handling %s option\n", progName, arg);
    			exit(2);
		} /*end switch */
	} /* end for */
}

void
semanticError(const char *msg)
{
	fprintf(stderr, "%s: %s\n", progName, msg);
	exit(1);
}

static void
badOptionCombo(const char *opt1, const char *opt2)
{
	struct argInfo *ai1 = findArg(opt1);
	struct argInfo *ai2 = findArg(opt2);
	assert(ai1 != NULL);
	assert(ai2 != NULL);
	if (ai1->argSeen && ai2->argSeen) {
		char msg[200];
		snprintf(msg, sizeof(msg), "Cannot specify %s with %s.", opt1, opt2);
		semanticError(msg);
	}
}

/*
 * The command line has been parsed.  Verify that the selected combination
 * of options, and the values of those options, make sense.  We finish most
 * of our initialization work as a by-product.
 */
void
checkSemantics()
{
	int status;

	/*
	 * If -help is specified, print the usage message.  It's an error
	 * if any other options are specified with -help.
	 */
	if (helpFlag) {
		usage();
		exit(somethingBesidesHelp ? 1 : 0);
	}
	else if (cfilterStr) {
		badOptionCombo("--compact", "--show-status");
		badOptionCombo("--compact", "--fix");
	}	
	else if (showstat) {
		badOptionCombo("--show-status", "--fix");
		badOptionCombo("--show-status", "--compr-bak");
	}
	else if (repairFlag) {
		badOptionCombo("--fix", "--force");	
	}
	else {
		usage();
	}			
}


void 
processCommand()
{
	int num_delrec;
	size_t delbytes;
	int flags = 0;
	
	if (logFilePath) {
		logfile = logFilePath;
	}
	else if (privFlag) {
		logfile = LOG_PRIVATE_PATH;
	}
	else {
		logfile = LOG_CURLOG_PATH;
	}
	if (cfilterStr) {
		if (forceFlag) {
			flags |= EVL_FORCE_COMPACT;
		}
		evl_compact_log(logfile, cfilterStr, flags);
	}
	else if (showstat) {
		
		if (evlGetNumDelRec(logfile, showstat, &num_delrec, &delbytes) == 0) {
			float kbytes = delbytes/1024;
			fprintf(stdout, "Number of records matching the filter is %d.\n"
							"Log file size would be reduced by %2.2f kbytes.\n", 
							 num_delrec, kbytes);	
		}	
	}
	else if (repairFlag) {
		evl_repair_log(logfile);
	}
		
}

void SIGTerm_handler()
{
	restoreLogfileFromBackup(logfile);
	exit(1);
}

int
main (int argc, char *argv[])
{
	static struct sigaction SigAllAction; /* Signal handler to terminate gracefully */
	static struct sigaction act;
	void SIGTerm_handler();
	
	uid_t uid = getuid();
	/* 
	 * Only root is allowed to run this tool, we check it here and also 
	 * install the binary with root permission only.
	 */
	if (uid != 0) {
		fprintf(stderr, "Only root is allowed to execute this utility.\n");
		exit(1);
	}
	/* Parse command line option */
	if (argc < 2) {
		usage();
	}
	
	processArgs(argc, argv);
	checkSemantics();

	/*
	 * Create new signal handler for SIGTERM. This will remove dr_socket and
	 * terminate gracefully.
	 */
	(void)memset(&SigAllAction, 0, sizeof(struct sigaction));
	SigAllAction.sa_handler = SIGTerm_handler;
	SigAllAction.sa_flags = 0;

	if (sigaction(SIGTERM, &SigAllAction, NULL) < 0){
		fprintf(stderr,
			"%s: WARNING - sigaction failed for SIGTERM.\n", argv[0]); 
	}
	if (sigaction(SIGINT, &SigAllAction, NULL) < 0){
		fprintf(stderr,
			"%s: WARNING - sigaction failed for new SIGINT.\n", argv[0]); 
	}
	if (sigaction(SIGHUP, &SigAllAction, NULL) < 0){
		fprintf(stderr,
			"%s: WARNING - sigaction failed for new SIGHUP.\n", argv[0]);
	}

	processCommand();
	return 0;
} 


